### PR TITLE
Fix gem uninstall behavior

### PR DESCRIPTION
**Pull request by bronzdoc**
# Description:
closes https://github.com/rubygems/rubygems/issues/2582

Oh boy... this was a tricky one for me...

This PR restores gem uninstall expected behavior which is to prompt a user to uninstall a specific version of a gem if there are multiple versions of the same gem installed i.e:
```
Select gem to uninstall:
 1. nitv-0.1.0
 2. nitv-0.1.1
 3. All versions
>
```
 but perseveres the new behavior introduced in https://github.com/rubygems/rubygems/pull/2466 which allow a user to specify the gem and version to uninstall by passing the gem in `gem:version` format.

Also i added a missing test in `test_gem_commands_uninstall_command.rb` that test gem selections when a gem is going to be uninstalled.
______________

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
Aha! So this is my regression... :sob:

Let me have a look :+1:.

File: lib/rubygems/commands/uninstall_command.rb [hunk 1]
```diff
@@ -148,21 +148,30 @@ def uninstall_all
...
+      if dep[:version] == Gem::Requirement.default
+        deps.map(&:name).uniq.each do |gem_name|
+          uninstall_gem(gem_name)
+        end
+      else
+        deps.each do |dep|
+          options[:version] = dep.version
+          uninstall_gem(dep.name)
+        end
+      end
```
Comment (deivid-rodriguez) [inline lib/rubygems/commands/uninstall_command.rb]: I think this nested loop might be creating some issues? See

### Currently

```
$ gem install rails:5.0.7.1 rails:5.1.6.1 rails:5.2.2 zeitwerk
Fetching rails-5.0.7.1.gem
Successfully installed rails-5.0.7.1
Successfully installed rails-5.1.6.1
Successfully installed rails-5.2.2
Fetching zeitwerk-1.3.1.gem
Successfully installed zeitwerk-1.3.1
4 gems installed

$ gem uninstall rails zeitwerk
Successfully uninstalled zeitwerk-1.3.1
Successfully uninstalled rails-5.0.7.1
Successfully uninstalled rails-5.1.6.1
Successfully uninstalled rails-5.2.2
```

### With this branch

```
$ RUBYOPT=-Ilib bin/gem install rails:5.0.7.1 rails:5.1.6.1 rails:5.2.2 zeitwerk
Fetching rails-5.0.7.1.gem
Successfully installed rails-5.0.7.1
Fetching rails-5.1.6.1.gem
Successfully installed rails-5.1.6.1
Fetching rails-5.2.2.gem
Successfully installed rails-5.2.2
Fetching zeitwerk-1.3.1.gem
Successfully installed zeitwerk-1.3.1
4 gems installed

$ RUBYOPT=-Ilib bin/gem uninstall rails zeitwerk
Successfully uninstalled zeitwerk-1.3.1

Select gem to uninstall:
 1. rails-5.0.7.1
 2. rails-5.1.6.1
 3. rails-5.2.2
 4. All versions
> 1
Successfully uninstalled rails-5.0.7.1

Select gem to uninstall:
 1. rails-5.1.6.1
 2. rails-5.2.2
 3. All versions
> 2
Successfully uninstalled rails-5.2.2
```

It's asking me the same question twice, which I already answered, so I don't know what to do.

Comment (bronzdoc) [reply lib/rubygems/commands/uninstall_command.rb]: ah nice catch, would add a test for that edge case

**Comment by hsbt** [comment]
@bronzdoc ping. I hope to merge this.

**Comment by bronzdoc** [comment]
@hsbt i'm on it :)

File: lib/rubygems/commands/uninstall_command.rb [hunk 2]
```diff
@@ -148,21 +148,30 @@ def uninstall_all
 
   def uninstall_specific
     deplist = Gem::DependencyList.new
+    original_gem_version = {}
+    gems_to_uninstall = {}
 
     get_all_gem_names_and_versions.each do |name, version|
-      requirement = Array(version || options[:version])
-      gem_specs = Gem::Specification.find_all_by_name(name, *requirement)
+      original_gem_version[name] = requirement = version || options[:version]
```
Comment (deivid-rodriguez) [inline lib/rubygems/commands/uninstall_command.rb]: The `requirement` local variable is now unused.

File: lib/rubygems/commands/uninstall_command.rb [hunk 3]
```diff
@@ -148,21 +148,30 @@ def uninstall_all
...
+
+      gem_specs = Gem::Specification.find_all_by_name(name, original_gem_version[name])
+
       say("Gem '#{name}' is not installed") if gem_specs.empty?
-      gem_specs.each do |spec|
-        deplist.add spec
-      end
+      gem_specs.each { |spec| deplist.add spec }
```
Comment (deivid-rodriguez) [inline lib/rubygems/commands/uninstall_command.rb]: I would leave the style untouched to make the diff of this bug fix simpler.

File: lib/rubygems/commands/uninstall_command.rb [hunk 4]
```diff
@@ -148,21 +148,30 @@ def uninstall_all
 
   def uninstall_specific
     deplist = Gem::DependencyList.new
+    original_gem_version = {}
+    gems_to_uninstall = {}
```
Comment (deivid-rodriguez) [inline lib/rubygems/commands/uninstall_command.rb]: This hash is only used in the second loop, so I would move it to right before it starts to make it more clear.

**Comment by bronzdoc** [comment]
> Do you plan to add the test about the edge case I mentioned 
Definitely! just need to figure out how to test it.

File: lib/rubygems/commands/uninstall_command.rb (hunk 3 above)
Comment (bronzdoc) [reply lib/rubygems/commands/uninstall_command.rb]: Fair 👍

**Comment by bronzdoc** [comment]
I think this is ready now @deivid-rodriguez @hsbt

File: test/rubygems/test_gem_commands_uninstall_command.rb [hunk 5]
```diff
@@ -192,6 +192,62 @@ def test_execute_with_version_specified_as_colon
     assert File.exist? File.join(@gemhome, 'bin', 'executable')
   end
 
+  def test_uninstall_selection
+    ui = Gem::MockGemUi.new "1\ny\n"
```
Comment (hsbt) [inline test/rubygems/test_gem_commands_uninstall_command.rb]: `y` is needless key type. I confirmed to pass the tests with `ui = Gem::MockGemUi.new "1\n"`

**Comment by deivid-rodriguez** [comment]
Awesome job!

**Comment by djberg96** [comment]
Does it still uninstall everything with the -a option?

**Comment by deivid-rodriguez** [comment]
I think so, yeah! :+1:

**Comment by deivid-rodriguez** [comment]
For reference, we're verifying `--all` in https://github.com/rubygems/rubygems/blob/fa6e547330e97b417ed11f262d68d03c57abeeda/test/rubygems/test_gem_commands_uninstall_command.rb#L21 and https://github.com/rubygems/rubygems/blob/fa6e547330e97b417ed11f262d68d03c57abeeda/test/rubygems/test_gem_commands_uninstall_command.rb#L245

So this should be good! :+1:

Thanks @bronzdoc for fixing my bug :)

@bundlerbot r+
